### PR TITLE
Add user properties C API

### PR DESCRIPTION
**Pull request by follitude**

### CODE REVIEW

File: librocksdb_sys/crocksdb/c.cc [hunk 1]
```diff
@@ -2188,6 +2195,103 @@ DB::RunManualCompaction
...
+                seq,
+                file_size);
+        return Status::OK();
+    }
+
+    virtual Status Finish(UserCollectedProperties *properties) override {
+        char **keys, **values;
+        int keys_count, value_count;
+        (*finish_)(state_, keys, &keys_count, values, &value_count);
+        if (keys == NULL || values == NULL) {
```
Comment (andelf) [inline librocksdb_sys/crocksdb/c.cc]: nullptr in C++

File: librocksdb_sys/crocksdb/c.cc [hunk 2]
```diff
@@ -2188,6 +2195,103 @@ DB::RunManualCompaction
...
+            return Status::OK();
+        }
+        assert(keys_count == value_count);
+        for (int i = 0; i < keys_count; i++) {
+            properties->insert(std::make_pair(std::string(keys[i]), std::string(values[i])));
+            free(keys[i]);
```
Comment (andelf) [inline librocksdb_sys/crocksdb/c.cc]: Bettter not free, if we use CString in rust, this segment of mem will be managed by rust allocator.

File: librocksdb_sys/crocksdb/c.cc [hunk 3]
```diff
@@ -2188,6 +2195,103 @@ DB::RunManualCompaction
...
+            free(values[i]);
+        }
+        // free the memory
+        free(keys);
```
Comment (andelf) [inline librocksdb_sys/crocksdb/c.cc]: Same reason as above.

File: librocksdb_sys/crocksdb/c.cc [hunk 4]
```diff
@@ -2188,6 +2195,103 @@ DB::RunManualCompaction
...
+
+void crocksdb_tablepropertiescollector_set_need_compact(
+        crocksdb_tablepropertiescollector_t *collector,
+        unsigned char whether_need_compact) {
+    collector->need_compact_ = whether_need_compact;
+}
+
+void crocksdb_options_add_table_properities_collector_factory(
+        crocksdb_options_t *opt, crocksdb_tablepropertiescollectorfactory_t *factory) {
+    opt->rep.table_properties_collector_factories.push_back(std::shared_ptr<TablePropertiesCollectorFactory>(factory));
```
Comment (andelf) [inline librocksdb_sys/crocksdb/c.cc]: use make_shared instead

File: librocksdb_sys/crocksdb/c.cc [hunk 5]
```diff
@@ -2188,6 +2195,103 @@ DB::RunManualCompaction
...
+    const char *(*name_)(void *);
+
+    virtual ~crocksdb_tablepropertiescollector_t() {
+        (*destructor_)(state_);
+    }
+
+    virtual Status AddUserKey(const Slice &key, const Slice &value,
+                              EntryType type, SequenceNumber seq,
+                              uint64_t file_size) override {
+        // For backwards-compatibility.
```
Comment (andelf) [inline librocksdb_sys/crocksdb/c.cc]: This comment is for rocksdb api, not here.

File: librocksdb_sys/crocksdb/c.cc [hunk 6]
```diff
@@ -2188,6 +2195,103 @@ DB::RunManualCompaction
...
+struct crocksdb_tablepropertiescollector_t : public TablePropertiesCollector {
+    void *state_;
+    void (*destructor_)(void *);
+    void (*add_userkey_)(
+            void *,
+            const char *key, size_t key_length,
+            const char *value, size_t value_length,
+            int entry_type, uint64_t seq,
+            uint64_t file_size);
+    void (*finish_)(void *, char **keys, int *key_count, char **value, int *value_count);
```
Comment (andelf) [inline librocksdb_sys/crocksdb/c.cc]: key_count and value_count always the same value, use pair_count or sth.

File: librocksdb_sys/crocksdb/c.cc [hunk 7]
```diff
@@ -2188,6 +2195,103 @@ DB::RunManualCompaction
...
+
+    virtual Status AddUserKey(const Slice &key, const Slice &value,
+                              EntryType type, SequenceNumber seq,
+                              uint64_t file_size) override {
+        // For backwards-compatibility.
+        (*add_userkey_)(
+                state_,
+                key.data(), key.size(),
+                value.data(), value.size(),
+                type,
```
Comment (andelf) [inline librocksdb_sys/crocksdb/c.cc]: Better use static_cast<int>()

File: librocksdb_sys/crocksdb/c.cc (hunk 3 above)
Comment (andelf) [inline librocksdb_sys/crocksdb/c.cc]: Normally we create a Vec<CString> then a Vec<*const c_char> in rust. V.as_ptr() returns *const *const c_char. And this is your "keys". Whole memory will be held by rust. Freed by RAII after calling crocksdb,

Comment (andelf) [inline librocksdb_sys/crocksdb/c.cc]: To avoid creating CString(which may be expensive), add const size_t* keys_lengths as extra parameter.

**Comment by andelf** [comment]
Will AddUserKey access prop kv pairs in its parameter? @zhangjinpeng1987

File: librocksdb_sys/crocksdb/c.cc [hunk 8]
```diff
@@ -2188,6 +2195,103 @@ DB::RunManualCompaction
...
+    crocksdb_tablepropertiescollector_t *(*create_table_properties_collector_)(
+            void *, crocksdb_tablepropertiescollectorfactorycontext_t *context);
+    const char *(*name_)(void *);
+
+    virtual ~crocksdb_tablepropertiescollectorfactory_t() { (*destructor_)(state_); }
+
+    virtual TablePropertiesCollector *CreateTablePropertiesCollector(
+            TablePropertiesCollectorFactory::Context context) {
+        crocksdb_tablepropertiescollectorfactorycontext_t ccontext;
+        ccontext.rep = context;
```
Comment (andelf) [inline librocksdb_sys/crocksdb/c.cc]: So an extra member access function is needed? For accessing context.column_family?

File: librocksdb_sys/crocksdb/c.cc (hunk 5 above)
Comment (siddontang) [inline librocksdb_sys/crocksdb/c.cc]: seem we don't need to support this function.
Why do you not support `Add` ?

File: librocksdb_sys/crocksdb/c.cc (hunk 2 above)
Comment (siddontang) [inline librocksdb_sys/crocksdb/c.cc]: better to define a free function like Merge interface does. 
But I guess maybe we can have a better way here.

File: librocksdb_sys/crocksdb/c.cc (hunk 6 above)
Comment (andelf) [inline librocksdb_sys/crocksdb/c.cc]: a remind, should be `char ***keys`

File: librocksdb_sys/crocksdb/c.cc [hunk 9]
```diff
@@ -2188,6 +2195,103 @@ DB::RunManualCompaction
 custom cache
 table_properties_collectors
 */
+struct crocksdb_tablepropertiescollectorfactorycontext_t {
```
Comment (zhangjinpeng1987) [inline librocksdb_sys/crocksdb/c.cc]: Move to line 131.

File: librocksdb_sys/crocksdb/c.cc (hunk 5 above)
Comment (zhangjinpeng1987) [inline librocksdb_sys/crocksdb/c.cc]: @siddontang `Add` is deprecate now, user defined collector should implement `AddUserKey`.

File: librocksdb_sys/crocksdb/c.cc [hunk 10]
```diff
@@ -2188,6 +2195,103 @@ DB::RunManualCompaction
...
+                type,
+                seq,
+                file_size);
+        return Status::OK();
+    }
+
+    virtual Status Finish(UserCollectedProperties *properties) override {
+        char **keys, **values;
+        int keys_count, value_count;
+        (*finish_)(state_, keys, &keys_count, values, &value_count);
```
Comment (zhangjinpeng1987) [inline librocksdb_sys/crocksdb/c.cc]: s/keys/&keys
s/values/&values

File: librocksdb_sys/crocksdb/c.cc (hunk 2 above)
Comment (andelf) [inline librocksdb_sys/crocksdb/c.cc]: 3 or 4 times of memcpy

**Comment by huachaohuang** [comment]
Should we add tests for these wrappers?

File: librocksdb_sys/crocksdb/c.cc (hunk 4 above)
Comment (siddontang) [inline librocksdb_sys/crocksdb/c.cc]: no, keep the same with other places.

File: librocksdb_sys/crocksdb/c.cc [hunk 11]
```diff
@@ -2188,6 +2195,103 @@ DB::RunManualCompaction
...
+        if (keys == NULL || values == NULL) {
+            return Status::OK();
+        }
+        assert(keys_count == value_count);
+        for (int i = 0; i < keys_count; i++) {
+            properties->insert(std::make_pair(std::string(keys[i]), std::string(values[i])));
```
Comment (huachaohuang) [inline librocksdb_sys/crocksdb/c.cc]: Try `properties->emplace()` to avoid copy.

**Comment by andelf** [comment]
Is this to be closed? for that #73 already handles all. @follitude

**Pull request closed without merging**